File: frontend/cyrene_ui/lib/models/chat_history.dart
class ChatHistory {
  final String id;
  final String sessionId;
  final String role; // 'user', 'assistant', 'system'
  final String content;
  final DateTime timestamp;
  final Map<String, dynamic>? metadata;
  final List<String>? attachments;
  final String? parentMessageId; // For threading/replies

  ChatHistory({
    required this.id,
    required this.sessionId,
    required this.role,
    required this.content,
    required this.timestamp,
    this.metadata,
    this.attachments,
    this.parentMessageId,
  });

  factory ChatHistory.fromJson(Map<String, dynamic> json) {
    return ChatHistory(
      id: json['id'],
      sessionId: json['session_id'],
      role: json['role'],
      content: json['content'],
      timestamp: DateTime.parse(json['timestamp']),
      metadata: json['metadata'],
      attachments: json['attachments'] != null
          ? List<String>.from(json['attachments'])
          : null,
      parentMessageId: json['parent_message_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'session_id': sessionId,
      'role': role,
      'content': content,
      'timestamp': timestamp.toIso8601String(),
      'metadata': metadata,
      'attachments': attachments,
      'parent_message_id': parentMessageId,
    };
  }
}

File: agent/db_core/models/chat_history.py
import uuid
from sqlalchemy import Column, Text, String, TIMESTAMP, Integer, ForeignKey, JSON
from sqlalchemy.dialects.postgresql import UUID, ARRAY, JSONB
from sqlalchemy.orm import relationship
from agent.db_core.core import Base

class ChatHistory(Base):
    __tablename__ = 'chat_history'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id = Column(UUID(as_uuid=True), ForeignKey('chat_sessions.id', ondelete='CASCADE'), nullable=False)
    role = Column(String(20), nullable=False)
    content = Column(Text, nullable=False)
    timestamp = Column(TIMESTAMP(timezone=True), server_default='NOW()')
    history_metadata = Column(JSONB, server_default="{}")
    attachments = Column(ARRAY(Text), server_default="{}")
    parent_message_id = Column(UUID(as_uuid=True), ForeignKey('chat_history.id'), nullable=True)
    token_count = Column(Integer, default=0)
    processing_time_ms = Column(Integer, default=0)

    session = relationship("ChatSession", back_populates="messages")
    parent = relationship("ChatHistory", remote_side=[id], backref="replies")
